doc/proposals/composable-integration-in sdk.md

```diff
@@ -0,0 +1,173 @@
+values that are computed dynamically when an application is deployed. For example, the admin URL for a Kafka
+deployment is known only after Kafka has been successfully deployed. If my CRD requires the URL, then
+rather that requiring a string value for that field, I can define it to be a reference to another object
+(perhaps the data is in the status field of the Kubernetes object managing Kafka).
+
+### User Story 2
+
+Often the yamls for a collection of resources require the same data in many places (e.g., a host name or port number).
+Rather than hardwiring the same value everywhere, I can define cross-resource references to a unique configmap that contains 
+the data. This makes it easier to change the data if needed, and can allow all resources to be configured
+dynamically (i.e. changes in the configmap are absorbed and resource that point to it are updated).
+
+
+### Implementation Details/Notes/Constraints (optional)
+
+The Composable SDK offers the following types.
+
+```golang
+type ObjectRef struct {
+	GetValueFrom ComposableGetValueFrom `json:"getValueFrom"`
+}
+
+type ComposableGetValueFrom struct {
+	Kind               string   `json:"kind"`
+	APIVersion         string   `json:"apiVersion,omitempty"`
+	Name               string   `json:"name,omitempty"`
+	Labels             []string `json:"labels,omitempty"`
+	Namespace          string   `json:"namespace,omitempty"`
+	Path               string   `json:"path"`
+	FormatTransformers []string `json:"format-transformers,omitempty"`
+}
+```
+
+An `ObjectRef` can be used to specify the type of any field of a CRD definition, allowing the value to be determined dynamically.
+For a detailed explanation of how to specify an object reference according to this schema, see [here](https://github.com/IBM/composable/blob/master/README.md#getvaluefrom-elements).
+
+The Composable SDK offers the following function for resolving the value of cross-resource references.
+
+```golang
+func ResolveObject(r client.Client, config *rest.Config, object interface{}, resolved interface{}, composableNamespace string) *ComposableError 
+```
+
+The function `ResolveObject` takes a Kubernetes client and configuration, and `object` to resolve, and a blank object
+`resolved` that will contain the result, as well as a namespace. The namespace is the default used when references
+do not specify one. This function will cast the result to the type of the `resolved` object, provided that
+appropriate data transformers have been included in the reference definitions (see [tutorial](https://github.com/IBM/composable/blob/master/sdk/docs/tutorial.md) for an example).
+
+The `ResolveObject` function uses caching for looking up object kinds, as well as for the objects themselves, in order
+to ensure that a consistent view of the data is obtained. If any data is not available at the time of the lookup,
+it returns an error. So this function either resolves the entire object or it doesn't -- there are no partial results.
```

Comment:
I think there's at least one interesting scenario that operator developer's using these object references would need to consider, and that is that the object resolves successfully and is created, and then at some point later, one of the references in a `GetValueFrom` field is deleted.

In that case, it looks like `ResolveObject()` will fail, but the caller may want to have some custom logic to handle the specific reason for the failure.

For example, if the error was a temporary connection issue, a caller would want to retry. But if the error is that a referenced object is not found, I probably still want to start retrying, but I might also want to delete the other objects that I created during a previous reconciliation that relied the resolved object that is now no longer resolving. Does that make sense?

Does the returned ComposableError.Error field have enough detail to allow this sort of decision making in the calling code?

Reply:
Currently ComposableError does not have this level of detail.